Using a View in CollectionOf combined with a View in Result doesn't seem to work correctly

Say we have the following scenario:

```golang
var _ = Service("test", func() {
	Method("testMethod", func() {
		Result(TestObj2, func() { View("objectView") })
		HTTP(func() {
			GET("/")
			Response(StatusOK)
		})
	})
})

var TestObj = ResultType("testObj", func() {
	Attributes(func() {
		Attribute("id", String)
		Attribute("someField", String)
	})

	View("default", func() {
		Attribute("id")
		Attribute("someField")
	})

	View("tiny", func() {
		Attribute("someField")
	})
})

var TestObj2 = ResultType("testObj2", func() {
	Attributes(func() {
		Attribute("id", String)
		Attribute("objects", CollectionOf(TestObj, func() {
			View("tiny")
		}))
	})

	View("default", func() {
		Attribute("id")
		Attribute("objects")
	})

	View("objectView", func() {
		Attribute("objects")
	})
})
```

The docs state that it should be possible to specify a View in Result, and also a View for a CollectionOf ResultType. However, combining the two seems to result in errors during generation:

```golang
/Users/tombrouws/Developer/test/test-backend/gen/test/service.go:38:1: expected type, found '}'
/Users/tombrouws/Developer/test/test-backend/gen/test/service.go:40:7: expected 'IDENT', found 'struct'
/Users/tombrouws/Developer/test/test-backend/gen/test/service.go:41:1: expected declaration, found '}'
/Users/tombrouws/Developer/test/test-backend/gen/test/service.go:46:2: expected declaration, found 'switch'
/Users/tombrouws/Developer/test/test-backend/gen/test/service.go:58:2: expected declaration, found 'switch'

========
Content:
// Code generated by goa v3.5.2, DO NOT EDIT.
//
// test service
//
// Command:
// $ goa gen test.nl/test/test-backend/design

package test

import (
        "context"
        "io"
        goa "goa.design/goa/v3/pkg"
        "goa.design/goa/v3/security"
        testviews "test.nl/test/test-backend/gen/test/views"
)


// Service is the test service interface.
type Service interface {
        // TestMethod implements testMethod.
                TestMethod(context.Context) (res *Testobj2, err error)
}

// ServiceName is the name of the service as defined in the design. This is the
// same value that is set in the endpoint request contexts under the ServiceKey
// key.
const ServiceName = "test"

// MethodNames lists the service method names as defined in the design. These
// are the same values that are set in the endpoint request contexts under the
// MethodKey key.
var MethodNames = [1]string{ "testMethod",  }
// Testobj2 is the result type of the test service testMethod method.
type Testobj2 struct {
        ID *string
        Objects *
}

type  struct {
}
// NewTestobj2 initializes result type Testobj2 from viewed result type
// Testobj2.
func NewTestobj2(vres *testviews.Testobj2, ) *Testobj2 {
        var res *Testobj2
        switch vres.View {
                case "default", "":
                                res = newTestobj2(vres.Projected)
                case "objectView":
                                res = newTestobj2ObjectView(vres.Projected)
        }
        return res
}
// NewViewedTestobj2 initializes viewed result type Testobj2 from result type
// Testobj2 using the given view.
func NewViewedTestobj2(res *Testobj2, view string, ) *testviews.Testobj2 {
        var vres *testviews.Testobj2
        switch view {
                case "default", "":
                                p := newTestobj2View(res)
                                vres = &testviews.Testobj2{Projected: p, View: "default" }
                case "objectView":
                                p := newTestobj2ViewObjectView(res)
                                vres = &testviews.Testobj2{Projected: p, View: "objectView" }
        }
        return vres
}
// newTestobj2 converts projected type Testobj2 to service type Testobj2.
func newTestobj2(vres *testviews.Testobj2View, ) *Testobj2 {
        res := &Testobj2{
ID: vres.ID,
}
                if vres.Objects != nil {
                        res.Objects = new(vres.Objects)
                }
        return res
}
// newTestobj2ObjectView converts projected type Testobj2 to service type
// Testobj2.
func newTestobj2ObjectView(vres *testviews.Testobj2View, ) *Testobj2 {
        res := &Testobj2{}
                if vres.Objects != nil {
                        res.Objects = new(vres.Objects)
                }
        return res
}
// newTestobj2View projects result type Testobj2 to projected type Testobj2View
// using the "default" view.
func newTestobj2View(res *Testobj2, ) *testviews.Testobj2View {
        vres := &testviews.Testobj2View{
ID: res.ID,
}
                if res.Objects != nil {
                        vres.Objects = newView(res.Objects)
                }
        return vres
}
// newTestobj2ViewObjectView projects result type Testobj2 to projected type
// Testobj2View using the "objectView" view.
func newTestobj2ViewObjectView(res *Testobj2, ) *testviews.Testobj2View {
        vres := &testviews.Testobj2View{}
                if res.Objects != nil {
                        vres.Objects = newView(res.Objects)
                }
        return vres
}
// new converts projected type  to service type .
func new(vres *testviews.View, ) * {
        res := &{}
        return res
}
// newView projects result type  to projected type View using the "default"
// view.
func newView(res *, ) *testviews.View {
        vres := &testviews.View{}
        return vres
}
```

Is this a bug, or am I doing something incorrectly?

--- a/expr/dup.go
+++ b/expr/dup.go
@@ -4,13 +4,7 @@ import "fmt"
 
 // Dup creates a copy the given data type.
 func Dup(d DataType) DataType {
-	res := newDupper().DupType(d)
-	if rt, ok := d.(*ResultTypeExpr); ok {
-		if Root.GeneratedResultType(rt.Identifier) != nil {
-			*Root.GeneratedTypes = append(*Root.GeneratedTypes, res.(*ResultTypeExpr))
-		}
-	}
-	return res
+	return newDupper().DupType(d)
 }
 
 // DupAtt creates a copy of the given attribute.
@@ -91,6 +85,16 @@ func (d *dupper) DupType(t DataType) DataType {
 		d.uts[actual.ID()] = dp
 		dupAtt := d.DupAttribute(actual.Attribute())
 		dp.SetAttribute(dupAtt)
+
+		// Make sure that if we are dupping a generated type we also put
+		// the dup in the generated type list so that it gets properly
+		// eval'd.
+		if rt, ok := dp.(*ResultTypeExpr); ok {
+			if Root.GeneratedResultType(rt.Identifier) != nil {
+				*Root.GeneratedTypes = append(*Root.GeneratedTypes, dp.(*ResultTypeExpr))
+			}
+		}
+
 		return dp
 	}
 	panic("unknown type " + fmt.Sprintf("%T", t))

